contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,253 @@
+kubectl's current extension point, unsuitable for credential rotation. It's easier
+to wrap `kubectl` so the tool is invoked on every command. For example, the following
+is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+client-go should be able to receive credentials by executing a command and reading
+that command's stdout.
+
+In fact, client-go already does this today. The GCP plugin can already be configured
+to [call a command](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/gcp/gcp.go#L228-L240)
+other than `gcloud`.
+
+## Plugin responsibilities
+
+Plugins are exec'd through client-go and print credentials to stdout. Errors are
+surfaced through stderr and a non-zero exit code. To use the bearer token
+`BEARER_TOKEN`, a plugin would print:
+
+```terminal
+$ ./kubectl-example-auth-plugin
+BEARER_TOKEN
+```
+
+Plugin output doesn't use versioned Kubernetes APIs (with kind and apiVersion)
+because we expect this plugin to interact with tools that don't natively understand
+Kubernetes. For example, the GCP auth provider execs [`gcloud auth`](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.9.0/plugin/pkg/client/auth/gcp/gcp.go#L230)
+and parses the output using JSONPath.
+
+To surface runtime-based information to the plugin, such as a request body for request
+signing, client-go will set the environment variable `KUBERNETES_EXEC_INFO` to a JSON
+serialized Kubernetes object when calling the plugin.
+
+```go
+type ExecInfo struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // No fields are proposed in this object yet.
+}
+```
+
+```
+KUBERNETES_EXEC_INFO='{"kind":"ExecInfo","apiVersion":"client-go.authentication.k8s.io/v1alpha1"}'
+```
+
+### Caching
+
+kubectl repeatedly [re-initializes transports](https://github.com/kubernetes/kubernetes/issues/37876)
+while client-go transports are long lived over many requests. As a result naive auth
+provider implementations that re-request credentials on every request have historically
+been slow.
+
+Plugins will be called on client-go initialization, and again when the API server returns
+a 401 HTTP status code indicating expired credentials. Credentials will otherwise be cached
```

Comment:
This is an unnecessary service disruption if a plugin knows when credentials will expire. Why would we not rotate proactively?